date-time/src/commonMain/kotlin/com/splendo/kaluga/datetime/timer/RecurringTimer.kt

```diff
@@ -0,0 +1,190 @@
+            /** Timer is paused. */
+            class Paused(
+                elapsedSoFar: Duration,
+                override val totalDuration: Duration
+            ) : NotRunning(elapsedSoFar), Timer.State.NotRunning.Paused {
+                fun start(
+                    interval: Duration,
+                    coroutineScope: CoroutineScope,
+                    finishCallback: suspend () -> Unit
+                ): Running = Running(
+                    elapsedSoFar = elapsed,
+                    totalDuration = totalDuration,
+                    interval = interval,
+                    coroutineScope = coroutineScope,
+                    finishCallback = finishCallback
+                )
+                internal fun finish(): Finished = Finished(totalDuration)
+            }
+
+            /** Timer is finished. */
+            class Finished(
+                override val totalDuration: Duration
+            ) : NotRunning(totalDuration), Timer.State.NotRunning.Finished
+        }
+
+        /** Timer is running. */
+        class Running(
+            elapsedSoFar: Duration,
+            override val totalDuration: Duration,
+            interval: Duration,
+            private val coroutineScope: CoroutineScope,
+            private val finishCallback: suspend () -> Unit
+        ) : State(), Timer.State.Running, HandleAfterNewStateIsSet<State>, HandleBeforeOldStateIsRemoved<State> {
+            override val elapsed: Flow<Duration> = tickProvider(
+                offset = elapsedSoFar,
+                max = totalDuration,
+                interval = interval
+            )
+            private lateinit var finishJob: Job
+
+            internal suspend fun pause(): NotRunning.Paused {
+                return NotRunning.Paused(elapsed.first(), totalDuration)
+            }
+            internal fun finish(): NotRunning.Finished = NotRunning.Finished(totalDuration)
+
+            override suspend fun beforeOldStateIsRemoved(oldState: State) {
+                finishJob = coroutineScope.launch {
+                    delay(totalDuration - elapsed.first())
+                    finishCallback()
+                }
+            }
+            override suspend fun afterNewStateIsSet(newState: State) {
+                finishJob.cancel()
+            }
+        }
+    }
+}
+
+private fun tickProvider(
+    offset: Duration,
+    max: Duration,
+    interval: Duration
+): Flow<Duration> {
+    val mark = TimeSource.Monotonic.markNow()
+
+    fun newElapsed() = offset + mark.elapsedNow()
+
+    return flow {
+        var expectedElapsed = offset
+        while (true) {
+            // obtain and deliver a new elapsed time value
+            val elapsed = newElapsed()
+            emit(elapsed.coerceAtMost(max))
+            // quit if the timer would reached the max on the next iteration
+            if (max - interval <= elapsed) break
```

Comment:
https://github.com/splendo/kaluga/pull/418#discussion_r759316679 also applies to this

Reply:
the last value would be produced by the finished state, and it would be `60_000`. the last value produced by the running state would be indeed close to `59_000`.

if done differently the `elapsed` flow would have duplicated `60_000` value (one from running and one from finished)

Reply:
so then `58_999` might create another tick but `59_001` not? 

Documenting and testing it matches documentation is still the way to go I think.

Reply:
removed custom handling of the last interval and added tests for time correction calculation